File: 18/sol.py
import timeit
from collections import defaultdict

import numpy as np

with open('input.txt') as f:
    lines = f.read().splitlines()
    lines = [l.split() for l in lines]
    # lines = [(x[0], int(x[1]), int(x[2][2:-1], 16)) for x in lines]
    lines = [(x[0], int(x[1]), x[2][2:-1]) for x in lines]

dirs = {
    'R': (0, 1),
    'D': (1, 0),
    'L': (0, -1),
    'U': (-1, 0),
    '0': (0, 1),
    '1': (1, 0),
    '2': (0, -1),
    '3': (-1, 0)
}


def PolyArea(x, y):
    x = np.asarray(x, dtype='int64')
    y = np.asarray(y, dtype='int64')
    return 0.5 * np.abs(np.dot(x, np.roll(y, 1)) - np.dot(y, np.roll(x, 1)))


def move(pos, dir, amount):
    dir = dirs[dir]
    return pos[0] + dir[0] * amount, pos[1] + dir[1] * amount


def f():
    pos = (0, 0)
    vertices = {(0, 0)}
    max_num = 0
    # vertices = defaultdict(set)
    xs = []
    ys = []

    for dir, _, colour in lines:
        # pt 2:
        dir = dirs[colour[-1]]
        amount = int(colour[:-1], 16)

        # dir = dirs[dir]
        for i in range(amount + 1):
            vertices.add((pos[0] + i * dir[0], pos[1] + i * dir[1]))
        pos = (pos[0] + amount * dir[0], pos[1] + amount * dir[1])
        max_num = max(max_num, pos[0], pos[1])
        xs.append(pos[0])
        ys.append(pos[1])

    print(max_num, len(vertices))
    area = PolyArea(xs, ys)
    print(f"Area: {area}")
    b = len(vertices)
    b2 = b // 2
    a1 = area + 1
    a2 = a1 - b2
    ans = a2 + b
    print(ans)
    # print((area + 1 - b2) + b)


if __name__ == '__main__':
    iters = 1
    x = timeit.timeit(lambda: f(), number=iters)
    print(f"{x: 0.2f} s")

File: 18/Sol.java
import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.*;

class Scratch {

    private record Pair(long y, long x) {
        Pair addMultiple(Pair other, long multiple) {
            return new Pair(y + other.y * multiple, x + other.x * multiple);
        }

        Pair copy() {
            return new Pair(y, x);
        }

        @Override
        public int hashCode() {
//            return Objects.hash(x, y);
            return (int)x * 31 + (int)y;
        }
    }

    private static Map<Character, Pair> dirs = Map.of(
            '0', new Pair(0, 1),
            '1', new Pair(1, 0),
            '2', new Pair(0, -1),
            '3', new Pair(-1, 0)
    );

    private static long area(List<Pair> points) {
        long area = 0;
        int j;
        for (int i = 0; i < points.size(); i++) {
            j = (i + 1) % points.size();
            area += points.get(i).x * points.get(j).y - points.get(j).x * points.get(i).y;
        }
        return area / 2;
    }

    public static void main(String[] args) throws IOException {
        var lines = Files.readAllLines(Path.of("C:\\Users\\phase\\PycharmProjects\\aoc2023\\18\\input.txt"));
        var pos = new Pair(0, 0);

        long start = System.currentTimeMillis();
        var vertices = new HashSet<Pair>(200000000, 0.75f); // resizing costs 2 seconds
        var points = new ArrayList<Pair>();


        for (String line : lines) {
            String[] parse = line.split(" ");
            Pair dir = dirs.get(parse[2].charAt(parse[2].length() - 2));
            long amount = Long.parseLong(parse[2].substring(2, 7), 16);
            for (long i = 0; i < amount; i++) {
                vertices.add(pos.addMultiple(dir, i));
            }
            pos = pos.addMultiple(dir, amount);
            points.add(pos.copy());
        }

        long area = area(points);
        long b = vertices.size();
        long b2 = b / 2;
        long a1 = area + 1;
        long a2 = a1 - b2;
        long ans = a2 + b;
        System.out.println("Time s: " + (System.currentTimeMillis() - start) / 1000f);
        System.out.println("Ans: " + ans);
    }
}